lisp/x/debugger/debugger.go

```diff
@@ -0,0 +1,989 @@
+	return profiler.GetFunNameFromFID(p.runtime, function.FunData().FID), function.FunData().Package
+}
+
+func (d *Debugger) End(function *lisp.LVal) {
+	if d.isEvaling {
+		return
+	}
+	d.logger.Debugf("End %v ", function.FunData().FID)
+	switch function.Type {
+	case lisp.LFun, lisp.LSymbol, lisp.LSExpr, lisp.LQSymbol:
+		d.decrementCallRef()
+	}
+	// no op for now except that we yield the CPU
+	runtime.Gosched()
+}
+
+func (d *Debugger) GetBreakpoint(id int) (*api.Breakpoint, error) {
+	d.Lock()
+	defer d.Unlock()
+	if bp, ok := d.breakpoints[id]; ok {
+		return bp, nil
+	}
+	return nil, errors.New("not found")
+}
+
+func (d *Debugger) GetBreakpointByName(name string) (*api.Breakpoint, error) {
+	d.Lock()
+	defer d.Unlock()
+	for _, v := range d.breakpoints {
+		if v.Name == name {
+			return v, nil
+		}
+	}
+	return nil, errors.New("not found")
+}
+
+func (d *Debugger) GetModules() []dap.Module {
+	modules := make([]dap.Module, 0)
+	for k := range d.runtime.Registry.Packages {
+		modules = append(modules, dap.Module{
+			Id:   k,
+			Name: k,
+		})
+	}
+	return modules
+}
+
+func (d *Debugger) Eval(text string) string {
+	d.Lock()
+	d.isEvaling = true
+	defer func() {
+		d.isEvaling = false
+		d.Unlock()
+	}()
+	d.logger.Debugf("Evaluating %s", text)
+	tEnv := lisp.NewEnv(nil)
+	tEnv.Runtime.Registry = d.runtime.Registry
+	tEnv.Runtime.Package = d.runtime.Package
+	tEnv.Runtime.Reader = d.runtime.Reader
+	tEnv.Runtime.Debugger = nil
+	v := tEnv.LoadString("eval", text)
+	return d.mapValue(v)
+}
+
+func (d *Debugger) GetDapStacktrace() []dap.StackFrame {
+	current := d.callRefs
+	out := make([]dap.StackFrame, 0)
+	for {
+		if current == nil {
+			break
+		}
+		d.logger.Debugf("Line %s in %s line %d col %d", current.name, current.file, current.line, current.col)
+		hint := "normal"
+		origin := ""
+		if current.file == "Unknown file" || current.file == "<native code>" {
+			hint = "deemphasize"
+			origin = "internal module"
+		}
+		out = append(out, dap.StackFrame{
+			Id:   0,
+			Name: current.name,
+			Source: dap.Source{
+				Name:             current.file,
+				Path:             current.path,
+				PresentationHint: hint,
+				Origin:           origin,
+			},
+			Line:             current.line,
+			Column:           current.col,
+			ModuleId:         current.packageName,
+			PresentationHint: hint,
+		})
+		current = current.prev
+	}
+	return out
+}
+
+func (d *Debugger) GetStacktrace(st *rpc2.StacktraceOut) {
+	d.logger.Debug("Returning STACK")
+	st.Locations = make([]api.Stackframe, 0)
+	for _, frame := range d.runtime.Stack.Frames {
+		var source = frame.Source
+		if source == nil {
+			source = &token.Location{
+				File: "Unknown file",
+				Path: "",
+				Pos:  0,
+				Line: 0,
+				Col:  0,
+			}
+		}
+		st.Locations = append(st.Locations, api.Stackframe{
+			Location: api.Location{
+				PC:   0,
+				File: fmt.Sprintf("%s", source.Path),
+				Line: source.Line,
+				Function: &api.Function{
+					Name_:     "f",
```

Comment:
Is it not possible to use LEnv.GetFunName here? I noticed this in delve.

```
(dlv) bt
0  0x0000000000000000 in f
   at approx.lisp:5
1  0x0000000000000000 in f
   at stream.lisp:5
2  0x0000000000000000 in f
   at :0
3  0x0000000000000000 in f
   at sicp.lisp:21
4  0x0000000000000000 in f
   at sicp.lisp:14
```

If the function name isn't available might it be better to just use the FID from the stack from?